src/malli/core.cljc

```diff
@@ -185,36 +175,33 @@
     [(first xs) (rest xs)]
     [nil xs]))
 
-(defn- -optional-entry? [[_ ?p]]
-  (boolean (and (map? ?p) (true? (:optional ?p)))))
-
-(defn- -optional-entry [[k ?p s :as entry] optional]
-  (if (map? ?p)
-    (cond
-      optional (update entry 1 assoc :optional true)
-      (= [:optional] (keys ?p)) [k s]
-      :else (update entry 1 dissoc :optional))
-    (if optional [k {:optional true} ?p] [k ?p])))
-
 (defn- -expand-key [[k ?p ?v] opts f]
-  (let [[p v] (if (map? ?p) [?p ?v] [nil ?p])]
+  (let [[p v] (if (or (nil? ?p) (map? ?p)) [?p ?v] [nil ?p])]
     [k p (f (schema v opts))]))
 
-(defn -parse-keys [children opts]
-  (let [entries (->> children (keep identity) (mapv #(-expand-key % opts identity)))]
-    {:required (->> entries (filter (comp not :optional second)) (mapv first))
-     :optional (->> entries (filter (comp :optional second)) (mapv first))
-     :keys (->> entries (mapv first))
-     :entries entries
-     :forms (mapv (fn [[k p v]]
-                    (let [v' (-form v)]
-                      (if p [k p v'] [k v']))) entries)}))
+(defn- -parse-map-entries [children opts]
+  (->> children (keep identity) (mapv #(-expand-key % opts identity))))
+
+(defn ^:no-doc map-entry-forms [entries]
+  (mapv (fn [[k p v]] (let [v' (-form v)] (if p [k p v'] [k v']))) entries))
+
+(defn ^:no-doc required-map-entry [[_ ?p]]
```

Comment:
Consider renaming `required-map-entry?` so it's obvious that it's a predicate. Also perhaps this should not be `no-doc`?

Reply:
renamed but left the `:no-doc` for now.